### PR TITLE
Refactor formatting of 'list' command

### PR DESCRIPTION
**Pull request by danieleades**
still further refactoring of the `dependencies.formatting` module.

not sure how you want to review/merge these. I guess you have the option of doing them individually, and I can rebase as you go, or as a single commit (this PR)

### CODE REVIEW

**Comment by danieleades** [comment]
any ideas why the coverage job is failing?

**Comment by foresterre** [comment]
> any ideas why the coverage job is failing?
> 
> scratch that, i see the issue about this. what changed? it used to work?

I've no idea why it doesn't work; it fails since https://github.com/foresterre/cargo-msrv/pull/227, possibly cargo-tarpaulin couldn't deal with integration [tests](https://github.com/foresterre/cargo-msrv/pull/227/files#diff-9f5acbbdbfbc8710ab5db5682583e25779610884c56efb4a167923e90b71a10bR94) where subprocesses return with non-zero exit codes; but I haven't found the root cause yet.

**Comment by foresterre** [comment]
bors try

**Comment by foresterre** [comment]
@danieleades Sorry for the review taking so long.

While I think this refactor contains some elements which make the code simpler, I intend to go to a smaller structs and traits kind of style (where behaviour is modeled using traits, and data encapsulated using small structs), like https://github.com/foresterre/cargo-msrv/pull/248/files, instead of a style with primarily free standing functions.

**Comment by danieleades** [comment]
Are you sure? This is the same logic, and essentially the same code, but it's nearly 100 lines of code shorter.

The original approach is *very* hard to follow

**Review by foresterre — APPROVED** [review]

**Comment by foresterre** [comment]
Would you please be so kind to rebase this PR one more time? 😅

**Comment by danieleades** [comment]
hmm. rebase turning out not to be trivial

**Comment by danieleades** [comment]
the tests are passing in CI, but they do not pass when i run cargo test locally (i also see the test failures on main)

**Comment by danieleades** [comment]
hmm. rebase turning out not to be trivial

**Comment by foresterre** [comment]
Would you mind if I rebased (or re-created, if it's hopeless) this PR?

**Comment by danieleades** [comment]
> Would you mind if I rebased (or re-created, if it's hopeless) this PR?
> 
> (I'll obviously keep you as author in git)

not at all! do whatever you like.

I did manage to rebase it in the end. Will need careful review though- i think it's ok, but i can't fully rule out that i messed something up in the rebase.

**Comment by foresterre** [comment]
I of course can't push to your branch ^^, I've created a new PR, from my repo, with your commits; The only change is an uppercase first letter in the commit message title, and I've squashed your commits. I also added one commit which fixes an accidental rename. The new PR is in #324.
